zeppelin-zengine/src/main/java/org/apache/zeppelin/search/SearchService.java

```diff
@@ -0,0 +1,352 @@
+  static final String ID_FIELD = "id";
+
+  Directory ramDirectory;
+  Analyzer analyzer;
+  IndexWriterConfig iwc;
+  IndexWriter writer;
+
+  public SearchService() {
+    ramDirectory = new RAMDirectory();
+    analyzer = new StandardAnalyzer();
+    iwc = new IndexWriterConfig(analyzer);
+    try {
+      writer = new IndexWriter(ramDirectory, iwc);
+    } catch (IOException e) {
+      LOG.error("Failed to reate new IndexWriter", e);
+    }
+  }
+
+  /**
+   * Full-text search in all the notebooks
+   *
+   * @param queryStr a query
+   * @return A list of matching paragraphs (id, text, snippet w/ highlight)
+   */
+  public List<Map<String, String>> query(String queryStr) {
+    if (null == ramDirectory) {
+      throw new IllegalStateException(
+          "Something went wrong on instance creation time, index dir is null");
+    }
+    List<Map<String, String>> result = Collections.emptyList();
+    try (IndexReader indexReader = DirectoryReader.open(ramDirectory)) {
+      IndexSearcher indexSearcher = new IndexSearcher(indexReader);
+      Analyzer analyzer = new StandardAnalyzer();
+      QueryParser parser = new QueryParser(SEARCH_FIELD, analyzer);
+
+      Query query = parser.parse(queryStr);
+      LOG.info("Searching for: " + query.toString(SEARCH_FIELD));
+
+      SimpleHTMLFormatter htmlFormatter = new SimpleHTMLFormatter();
+      Highlighter highlighter = new Highlighter(htmlFormatter, new QueryScorer(query));
+
+      result = doSearch(indexSearcher, query, analyzer, highlighter);
+      indexReader.close();
+    } catch (IOException e) {
+      LOG.error("Failed to open index dir {}, make sure indexing finished OK", ramDirectory, e);
+    } catch (ParseException e) {
+      LOG.error("Failed to parse query " + queryStr, e);
+    }
+    return result;
+  }
+
+  private List<Map<String, String>> doSearch(IndexSearcher searcher, Query query,
+      Analyzer analyzer, Highlighter highlighter) {
+    List<Map<String, String>> matchingParagraphs = Lists.newArrayList();
+    ScoreDoc[] hits;
+    try {
+      hits = searcher.search(query, 20).scoreDocs;
+      for (int i = 0; i < hits.length; i++) {
+        LOG.info("doc={} score={}", hits[i].doc, hits[i].score);
+
+        int id = hits[i].doc;
+        Document doc = searcher.doc(id);
+        String path = doc.get(ID_FIELD);
+        if (path != null) {
+          LOG.info((i + 1) + ". " + path);
+          String title = doc.get("title");
+          if (title != null) {
+            LOG.info("   Title: {}", doc.get("title"));
+          }
+
+          String text = doc.get(SEARCH_FIELD);
+          TokenStream tokenStream = TokenSources.getTokenStream(searcher.getIndexReader(), id,
+              SEARCH_FIELD, analyzer);
+          TextFragment[] frag = highlighter.getBestTextFragments(tokenStream, text, true, 3);
+          // TODO(bzz): remove this as too verbose
+          LOG.info("    {} fragments found for query '{}' in '{}'", frag.length, query, text);
```

Comment:
We might not want to print contents of notebook in a log file. Either use LOG.debug or remove 'text' from the log.